# Bag of Words Representation/Model

<b>The Bag of Words (BoW)</b> is a simple and commonly used model in natural language processing (NLP) to represent text data for machine learning. It transforms text into a format that algorithms can work with, particularly by counting the frequency of words in a given set of documents.

Key Concepts of Bag of Words:

- <b>Vocabulary:</b> This refers to the set of unique words that appear in the corpus (collection of documents). Each word in the vocabulary becomes a column in the resulting matrix.

- <b>Document Representation:</b> Each document in the corpus is represented as a vector, where each element corresponds to the frequency of a word from the vocabulary. If a word does not appear in the document, its frequency will be zero.

- <b>Order Ignorance:</b> The Bag of Words model ignores the order of words in the text. It only counts the occurrence of each word without considering the context or sequence in which the words appear.

In [1]:
# import required libraries
import pandas as pd
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import CountVectorizer

### Basic Bag of Words Example

In [2]:
documents = ["Gangs of Wasseypur is a great movie.", "The success of a movie depends on the performance of the actors.", "There are no new movies releasing this week."]
print(documents)

['Gangs of Wasseypur is a great movie.', 'The success of a movie depends on the performance of the actors.', 'There are no new movies releasing this week.']


#### Let us define a method for preprocessing the document/text. Steps are

- convert the text to lower/upper case (preferred lower case) to make the words case insensitive
- tokenise the text into words
- remove the stop words
- re-join the tokenised words into a sentence

In [3]:
def preprocess(document):

    # convert document text to lower
    document = document.lower()

    # tokenise the text into words
    words = word_tokenize(document)

    # remove the stop words
    words = [word for word in words if word not in stopwords.words('english')]

    # rejoin the words to make sentence without stop words
    document = ' '.join(words)

    return document

In [4]:
# Let is pre-process the documents
documents = [preprocess(document) for document in documents]
print(documents)

['gangs wasseypur great movie .', 'success movie depends performance actors .', 'new movies releasing week .']


#### Creating bag of words using count vectorizer function

In [5]:
vectorizer = CountVectorizer()
bow_model = vectorizer.fit_transform(documents)
print(bow_model)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 13 stored elements and shape (3, 12)>
  Coords	Values
  (0, 2)	1
  (0, 10)	1
  (0, 3)	1
  (0, 4)	1
  (1, 4)	1
  (1, 9)	1
  (1, 1)	1
  (1, 7)	1
  (1, 0)	1
  (2, 6)	1
  (2, 5)	1
  (2, 8)	1
  (2, 11)	1


In [6]:
# print the full sparse matrix
print(bow_model.toarray())

[[0 0 1 1 1 0 0 0 0 0 1 0]
 [1 1 0 0 1 0 0 1 0 1 0 0]
 [0 0 0 0 0 1 1 0 1 0 0 1]]


In [7]:
print(bow_model.shape)

(3, 12)


In [8]:
print(vectorizer.get_feature_names_out())

['actors' 'depends' 'gangs' 'great' 'movie' 'movies' 'new' 'performance'
 'releasing' 'success' 'wasseypur' 'week']


### Create Bag of Words for Spam-Ham Dataset 

In [9]:
# read the file
spam_ham_text = pd.read_csv('SMSSpamCollection.txt', delimiter='\t', header=None)

In [10]:
spam_ham_text.head()

0                                                  1
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [11]:
spam_ham_text.columns = ['label','message']

In [12]:
spam_ham_text.head()

label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [13]:
spam_ham_text.shape

(5572, 2)

In [14]:
spam_ham_text.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   label    5572 non-null   object
 1   message  5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


In [15]:
# taking only first 50 records from the entire data frame
spam_ham_text = spam_ham_text.iloc[0:50,:]

In [16]:
spam_ham_text.head()

label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [17]:
messages = spam_ham_text.message

In [18]:
type(messages)

pandas.core.series.Series

In [19]:
# lets convert the messages to list for preprocessing
messages = [message for message in messages]

In [20]:
type(messages)

list

In [21]:
# Let us pre-process the messages
messages = [preprocess(message) for message in messages]
print(messages)

['go jurong point , crazy .. available bugis n great world la e buffet ... cine got amore wat ...', 'ok lar ... joking wif u oni ...', "free entry 2 wkly comp win fa cup final tkts 21st may 2005. text fa 87121 receive entry question ( std txt rate ) & c 's apply 08452810075over18 's", 'u dun say early hor ... u c already say ...', "nah n't think goes usf , lives around though", "freemsg hey darling 's 3 week 's word back ! 'd like fun still ? tb ok ! xxx std chgs send , £1.50 rcv", 'even brother like speak . treat like aids patent .', "per request 'melle melle ( oru minnaminunginte nurungu vettam ) ' set callertune callers . press * 9 copy friends callertune", 'winner ! ! valued network customer selected receivea £900 prize reward ! claim call 09061701461. claim code kl341 . valid 12 hours .', 'mobile 11 months ? u r entitled update latest colour mobiles camera free ! call mobile update co free 08002986030', "'m gon na home soon n't want talk stuff anymore tonight , k ? 've cried enoug

In [22]:
# vectorize the messages
vectorizer = CountVectorizer()
bow_spam_ham = vectorizer.fit_transform(messages)

In [23]:
print(bow_spam_ham)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 467 stored elements and shape (50, 381)>
  Coords	Values
  (0, 139)	1
  (0, 171)	1
  (0, 244)	1
  (0, 87)	1
  (0, 44)	1
  (0, 52)	1
  (0, 147)	1
  (0, 370)	1
  (0, 176)	1
  (0, 51)	1
  (0, 71)	1
  (0, 144)	1
  (0, 34)	1
  (0, 351)	1
  (1, 231)	1
  (1, 177)	1
  (1, 170)	1
  (1, 360)	1
  (1, 232)	1
  (2, 131)	1
  (2, 116)	2
  (2, 363)	1
  (2, 80)	1
  (2, 361)	1
  (2, 118)	2
  :	:
  (47, 142)	1
  (47, 37)	1
  (47, 120)	1
  (48, 42)	1
  (48, 376)	1
  (48, 157)	1
  (48, 333)	1
  (48, 55)	1
  (48, 85)	1
  (48, 201)	1
  (48, 43)	1
  (48, 47)	1
  (49, 139)	1
  (49, 117)	1
  (49, 348)	1
  (49, 199)	1
  (49, 174)	1
  (49, 294)	1
  (49, 159)	1
  (49, 172)	1
  (49, 309)	1
  (49, 355)	2
  (49, 297)	1
  (49, 160)	1
  (49, 298)	1


In [24]:
print(bow_spam_ham.toarray())

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [25]:
print(bow_spam_ham.shape)
print(vectorizer.get_feature_names_out())

(50, 381)
['000' '07732584351' '08000930705' '08002986030' '08452810075over18'
 '09061701461' '100' '11' '12' '150p' '16' '20' '2005' '21st' '2nd'
 '4403ldnw1a7rw18' '4txt' '50' '6days' '81010' '87077' '87121' '87575'
 '8am' '900' 'abiola' 'actin' 'aft' 'ahead' 'ahhh' 'aids' 'already'
 'alright' 'always' 'amore' 'amp' 'anymore' 'anything' 'apologetic'
 'apply' 'arabian' 'ard' 'around' 'ask' 'available' 'back' 'badly' 'bit'
 'blessing' 'breather' 'brother' 'buffet' 'bugis' 'burns' 'bus' 'ca'
 'call' 'callers' 'callertune' 'calls' 'camcorder' 'camera' 'car' 'cash'
 'catch' 'caught' 'chances' 'charged' 'cheers' 'chgs' 'child' 'cine'
 'claim' 'clear' 'click' 'co' 'code' 'colour' 'com' 'comin' 'comp'
 'confirm' 'convincing' 'copy' 'cost' 'could' 'crave' 'crazy' 'credit'
 'cried' 'csh11' 'cup' 'cuppa' 'customer' 'da' 'darling' 'date' 'day'
 'dbuk' 'decide' 'decided' 'delivery' 'dinner' 'done' 'dont' 'dun' 'early'
 'eat' 'eating' 'eg' 'egg' 'eh' 'endowed' 'england' 'enough' 'entitled'
 'entry

#### A lot of duplicate tokens such as 'win'and 'winner'; 'reply' and 'replying'; 'want' and 'wanted' etc. 

#### Let us add lemmatisation and stemming step in above pre-processing

In [26]:
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer

stemmer = PorterStemmer()
wordnet_lemmatizer = WordNetLemmatizer()

def preprocess(document, stem=True):

    # convert document text to lower
    document = document.lower()

    # tokenise the text into words
    words = word_tokenize(document)

    # remove the stop words
    words = [word for word in words if word not in stopwords.words('english')]

    if stem:
        words = [stemmer.stem(word) for word in words]
    else:
        words = [wordnet_lemmatizer.lemmatize(word, pos='v') for word in words]
    
    # rejoin the words to make sentence without stop words
    document = ' '.join(words)

    return document

### Bag of words model on stemmed messages

In [27]:
# stem messages
messages = [preprocess(message, stem=True) for message in messages]

# bag of words model
vectorizer = CountVectorizer()
bow_model = vectorizer.fit_transform(messages)

In [28]:
# Dataframe
pd.DataFrame(bow_model.toarray(), columns = vectorizer.get_feature_names_out())

000  07732584351  08000930705  08002986030  08452810075over18  \
0     0            0            0            0                  0   
1     0            0            0            0                  0   
2     0            0            0            0                  1   
3     0            0            0            0                  0   
4     0            0            0            0                  0   
5     0            0            0            0                  0   
6     0            0            0            0                  0   
7     0            0            0            0                  0   
8     0            0            0            0                  0   
9     0            0            0            1                  0   
10    0            0            0            0                  0   
11    1            0            0            0                  0   
12    1            0            0            0                  0   
13    0            0            0            0                  0   
14    0            0            0            0                  0   
15    0            0            0            0                  0   
16    0            0            0            0                  0   
17    0            0            0            0                  0   
18    0            0            0            0                  0   
19    0            0            0            0                  0   
20    0            0            0            0                  0   
21    0            0            0            0                  0   
22    0            0            0            0                  0   
23    0            0            0            0                  0   
24    0            0            0            0                  0   
25    0            0            0            0                  0   
26    0            0            0            0                  0   
27    0            0            0            0                  0   
28    0            0            0            0                  0   
29    0            0            0            0                  0   
30    0            0            0            0                  0   
31    0            0            0            0                  0   
32    0            0            0            0                  0   
33    0            0            0            0                  0   
34    0            0            0            0                  0   
35    0            0            0            0                  0   
36    0            0            0            0                  0   
37    0            0            0            0                  0   
38    0            0            0            0                  0   
39    0            0            0            0                  0   
40    0            0            0            0                  0   
41    0            0            0            0                  0   
42    0            1            1            0                  0   
43    0            0            0            0                  0   
44    0            0            0            0                  0   
45    0            0            0            0                  0   
46    0            0            0            0                  0   
47    0            0            0            0                  0   
48    0            0            0            0                  0   
49    0            0            0            0                  0   

    09061701461  100  11  12  150p  ...  worri  www  xuhui  xxx  \
0             0    0   0   0     0  ...      0    0      0    0   
1             0    0   0   0     0  ...      0    0      0    0   
2             0    0   0   0     0  ...      0    0      0    0   
3             0    0   0   0     0  ...      0    0      0    0   
4             0    0   0   0     0  ...      0    0      0    0   
5             0    0   0   0     0  ...      0    0      0    1   
6             0

In [ ]:
# token names
vectorizer.get_feature_names_out()

#### So we see 357 tokens with stemming as comapared to 381 tokens without stemming

### Bag of words model on lemmatized messages

In [30]:
# stem messages
messages = [preprocess(message, stem=False) for message in messages]

# bag of words model
vectorizer = CountVectorizer()
bow_model = vectorizer.fit_transform(messages)

In [31]:
# Dataframe
pd.DataFrame(bow_model.toarray(), columns = vectorizer.get_feature_names_out())

000  07732584351  08000930705  08002986030  08452810075over18  \
0     0            0            0            0                  0   
1     0            0            0            0                  0   
2     0            0            0            0                  1   
3     0            0            0            0                  0   
4     0            0            0            0                  0   
5     0            0            0            0                  0   
6     0            0            0            0                  0   
7     0            0            0            0                  0   
8     0            0            0            0                  0   
9     0            0            0            1                  0   
10    0            0            0            0                  0   
11    1            0            0            0                  0   
12    1            0            0            0                  0   
13    0            0            0            0                  0   
14    0            0            0            0                  0   
15    0            0            0            0                  0   
16    0            0            0            0                  0   
17    0            0            0            0                  0   
18    0            0            0            0                  0   
19    0            0            0            0                  0   
20    0            0            0            0                  0   
21    0            0            0            0                  0   
22    0            0            0            0                  0   
23    0            0            0            0                  0   
24    0            0            0            0                  0   
25    0            0            0            0                  0   
26    0            0            0            0                  0   
27    0            0            0            0                  0   
28    0            0            0            0                  0   
29    0            0            0            0                  0   
30    0            0            0            0                  0   
31    0            0            0            0                  0   
32    0            0            0            0                  0   
33    0            0            0            0                  0   
34    0            0            0            0                  0   
35    0            0            0            0                  0   
36    0            0            0            0                  0   
37    0            0            0            0                  0   
38    0            0            0            0                  0   
39    0            0            0            0                  0   
40    0            0            0            0                  0   
41    0            0            0            0                  0   
42    0            1            1            0                  0   
43    0            0            0            0                  0   
44    0            0            0            0                  0   
45    0            0            0            0                  0   
46    0            0            0            0                  0   
47    0            0            0            0                  0   
48    0            0            0            0                  0   
49    0            0            0            0                  0   

    09061701461  100  11  12  150p  ...  worri  www  xuhui  xxx  \
0             0    0   0   0     0  ...      0    0      0    0   
1             0    0   0   0     0  ...      0    0      0    0   
2             0    0   0   0     0  ...      0    0      0    0   
3             0    0   0   0     0  ...      0    0      0    0   
4             0    0   0   0     0  ...      0    0      0    0   
5             0    0   0   0     0  ...      0    0      0    1   
6             0

In [32]:
# token names
vectorizer.get_feature_names_out()

array(['000', '07732584351', '08000930705', '08002986030',
       '08452810075over18', '09061701461', '100', '11', '12', '150p',
       '16', '20', '2005', '21st', '2nd', '4403ldnw1a7rw18', '4txt', '50',
       '6day', '81010', '87077', '87121', '87575', '8am', '900', 'abiola',
       'actin', 'aft', 'ahead', 'ahhh', 'aid', 'alreadi', 'alright',
       'alway', 'amor', 'amp', 'anymor', 'anyth', 'apologet', 'appli',
       'arabian', 'ard', 'around', 'ask', 'avail', 'back', 'badli',
       'bite', 'bless', 'breather', 'brother', 'bu', 'buffet', 'bugi',
       'burn', 'ca', 'call', 'caller', 'callertun', 'camcord', 'camera',
       'car', 'cash', 'catch', 'chanc', 'charg', 'cheer', 'chg', 'child',
       'cine', 'claim', 'clear', 'click', 'co', 'code', 'colour', 'com',
       'comin', 'comp', 'confirm', 'convinc', 'copi', 'cost', 'could',
       'crave', 'crazi', 'credit', 'cri', 'csh11', 'cup', 'cuppa',
       'custom', 'da', 'darl', 'date', 'day', 'dbuk', 'decid', 'deliveri',
       'd

### With lemmatisation there ate 355 tokens